src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -587,25 +589,25 @@ sizeInPages = VU.length . ciPrimary
 fromList :: Int -> Int -> [Append] -> CompactIndex
 fromList rfprec maxcsize apps = runST $ do
     mci <- new rfprec maxcsize
-    cs <- mapM (`append` mci) apps
-    (c, fc) <- unsafeEnd mci
-    pure (fromChunks (concat cs ++ toList c) fc)
+    mapM_ (`append` mci) apps
+    (_, fc, primary) <- unsafeEnd mci
+    pure (fromChunks primary fc)
 
 -- | One-shot construction using only 'appendSingle'.
 fromListSingles :: Int -> Int -> [(SerialisedKey, SerialisedKey)] -> CompactIndex
 fromListSingles rfprec maxcsize apps = runST $ do
     mci <- new rfprec maxcsize
-    cs <- mapM (`appendSingle` mci) apps
-    (c, fc) <- unsafeEnd mci
-    pure (fromChunks (catMaybes cs ++ toList c) fc)
+    mapM_ (`appendSingle` mci) apps
+    (_, fc, primary) <- unsafeEnd mci
+    pure (fromChunks primary fc)
 
 -- | Feed in 'Chunk's in the same order that they were yielded from incremental
 -- construction using 'MCompactIndex'.
-fromChunks :: [Chunk] -> FinalChunk -> CompactIndex
-fromChunks cs FinalChunk{..} = CompactIndex {
+fromChunks :: PrimaryArray -> FinalChunk -> CompactIndex
+fromChunks (PrimaryArray primary) FinalChunk{..} = CompactIndex {
```

Comment:
See my earlier comment -- let's make a test-only function, either making it explicit using some comments or by moving it to a test module